cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -872,15 +872,15 @@ private predicate simple_comparison_eq(
 }
 
 private predicate complex_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
```

Comment:
Is there a reason for not using the newly introduced `CompareValueNumber` here?

Reply:
It actually does matter because of the `LogicalNotInstruction` case:
```ql
exists(AbstractValue dual, LogicalNotInstruction logicalNot |
  value = dual.getDualValue() and
  logicalNot = test.getAnInstruction() and
  compares_eq(valueNumber(logicalNot.getUnary()), left, right, k, areEqual, dual)
)
```
here we don't require the value number to be a `CompareValueNumber`, but instead we require it to be something of the form `!x`, where the `x` is then used recursively.

I'll see if I can come up with a testcase where this is important 🤔